### PR TITLE
Real life hotfixes

### PR DESCRIPTION
**Pull request by StijnVrolijk**
When using this bundle in a real life project I came across a few issues.

### CODE REVIEW

**Comment by bjorvack** [comment]
@StijnVrolijk tests failen nog

File: User/InMemoryUserRepository.php [hunk 1]
```diff
@@ -92,22 +92,22 @@ public function getSupportedClass()
     /**
      * {@inheritdoc}
      */
-    public function add(UserInterface $user)
+    public function add(User $user)
```
Comment (carakas) [inline User/InMemoryUserRepository.php]: why did you swap the interface for a User ? or is this also an interface ?

File: User/InMemoryUserRepository.php [hunk 2]
```diff
@@ -27,15 +27,15 @@ public function __construct()
 
         $this->users[] = $user;
 
-        $passwordResetUser = new User(
+        $passwordResetUser = new UserWithPassword(
```
Comment (carakas) [inline User/InMemoryUserRepository.php]: So there are also users without passwords ?

File: User/InMemoryUserRepository.php [hunk 3]
```diff
@@ -92,22 +92,22 @@ public function getSupportedClass()
...
+    public function add(User $user)
     {
         $this->users[] = $user;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function save(UserInterface $user)
+    public function save(User $user)
```
Comment (carakas) [inline User/InMemoryUserRepository.php]: shouldn't you save the user in the array here ?

File: User/UserWithPassword.php [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+     */
+    public function getUsername()
+    {
+        return $this->username;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function eraseCredentials()
```
Comment (carakas) [inline User/UserWithPassword.php]: shouldn't this do something ?

File: User/UserWithPassword.php [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    /**
+     * @return string
+     */
+    public function getEmail()
+    {
+        return $this->email;
+    }
+
+    /**
+     * @param $password
```
Comment (carakas) [inline User/UserWithPassword.php]: string :)

File: User/InMemoryUserRepository.php (hunk 2 above)
Comment (StijnVrolijk) [inline User/InMemoryUserRepository.php]: There used to be... I can imagine some situations where a user wouldn't have a password. Plus I think it reads better when you can say `class Admin extends UserWithPassword`. It's just an extra confirmation

File: User/InMemoryUserRepository.php (hunk 3 above)
Comment (StijnVrolijk) [inline User/InMemoryUserRepository.php]: save is the method used when updating but this happens automagically when editing the original object (at least in memory)

File: User/UserWithPassword.php (hunk 4 above)
Comment (StijnVrolijk) [inline User/UserWithPassword.php]: I'm passing this along to @bjorvack , he wrote this part

File: User/UserWithPassword.php (hunk 5 above)
Comment (StijnVrolijk) [inline User/UserWithPassword.php]: well spotted, sir. 7c9bf20

File: User/InMemoryUserRepository.php (hunk 2 above)
Comment (carakas) [inline User/InMemoryUserRepository.php]: ok, I can follow that logic :)

File: User/InMemoryUserRepository.php (hunk 3 above)
Comment (carakas) [inline User/InMemoryUserRepository.php]: ok

Comment (carakas) [inline User/InMemoryUserRepository.php]: maybe just add a comment stating that here

File: User/UserWithPassword.php (hunk 5 above)
Comment (carakas) [inline User/UserWithPassword.php]: 👍

File: User/UserWithPassword.php (hunk 4 above)
Comment (bjorvack) [inline User/UserWithPassword.php]: I'm passing this to @WouterSioen, he wrote this part

Comment (carakas) [inline User/UserWithPassword.php]: looks like a hot patato

File: User/InMemoryUserRepository.php (hunk 3 above)
Comment (StijnVrolijk) [inline User/InMemoryUserRepository.php]: I don't think that's necessary but i did it anyway, just for you 85d8c2f

File: User/UserWithPassword.php (hunk 4 above)
Comment (WouterSioen) [inline User/UserWithPassword.php]: I wrote this for the user without a password, in which case it shouldn't do a thing.

When a user has a password, this one should erase the plainPassword from the entity, only keeping the encoded version.

Comment (StijnVrolijk) [inline User/UserWithPassword.php]: that reminds me.. the password is still saved in plaintext

Comment (WouterSioen) [inline User/UserWithPassword.php]: oops :D

File: DataTransferObject/UserDataTransferObject.php [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace SumoCoders\FrameworkMultiUserBundle\DataTransferObject;
+
+use SumoCoders\FrameworkMultiUserBundle\User\User;
+
+interface UserDataTransferObject
+{
+    public static function fromUser(User $user);
+    public function getEntity();
```
Comment (WouterSioen) [inline DataTransferObject/UserDataTransferObject.php]: This kind of makes them factories instead of Data Transfer Objects.

Comment (StijnVrolijk) [inline DataTransferObject/UserDataTransferObject.php]: Doesn't a factory process it's data immediatly? I see these objects more as a temporary storage of data

Comment (WouterSioen) [inline DataTransferObject/UserDataTransferObject.php]: well, the `getEntity` makes an object based on some parameters, which makes it a factory method.

File: User/UserWithPassword.php (hunk 4 above)
Comment (StijnVrolijk) [inline User/UserWithPassword.php]: there, i fixed it

File: DataTransferObject/UserDataTransferObject.php (hunk 6 above)
Comment (StijnVrolijk) [inline DataTransferObject/UserDataTransferObject.php]: @WouterSioen should I rename them to factories?

**Comment by WouterSioen** [comment]
Stijn: breaking tests since 1989

**Comment by StijnVrolijk** [comment]
There, I fixed it!

File: DataTransferObject/UserDataTransferObject.php (hunk 6 above)
Comment (WouterSioen) [inline DataTransferObject/UserDataTransferObject.php]: I don't know :( They are both. Can't decide. I'll think about it.

**Comment by StijnVrolijk** [comment]
This seems to be it for now, guys. What do you think?

**Comment by StijnVrolijk** [comment]
OK, I lied before but I swear I'm not lying right now.

**Comment by StijnVrolijk** [comment]
@sumocoders/php-team please let me know what you think, I need this fairly urgently :)

**Comment by carakas** [comment]
:shipit:

**Comment by bjorvack** [comment]
:shipit:

**Pull request merged**